src/lib/Libifl/tcp_dis.c

```diff
@@ -109,46 +109,35 @@ tcp_recv(int fd, void *data, int len)
 	int torecv = len;
 	char *pb = (char *)data;
 	int amt = 0;
-#ifdef WIN32
 	fd_set readset;
 	struct timeval timeout;
-#else
-	struct pollfd pollfds[1];
-	int timeout;
-#endif
 
-	/*
-	 * we don't want to be locked out by an attack on the port to
-	 * deny service, so we time out the read, the network had better
-	 * deliver promptly
-	 */
-	do {
-#ifdef WIN32
-		timeout.tv_sec = (long) pbs_tcp_timeout;
-		timeout.tv_usec = 0;
+	timeout.tv_sec = (time_t) pbs_tcp_timeout;
+	timeout.tv_usec = 0;
+
+	while (torecv > 0) {
 		FD_ZERO(&readset);
 		FD_SET((unsigned int)fd, &readset);
-		i = select(FD_SETSIZE, &readset, NULL, NULL, &timeout);
-#else
-		timeout = pbs_tcp_timeout;
-		pollfds[0].fd = fd;
-		pollfds[0].events = POLLIN;
-		pollfds[0].revents = 0;
-		i = poll(pollfds, 1, timeout * 1000);
-#endif
-		if (pbs_tcp_interrupt)
-			break;
-	}
+
+		/*
+		 * we don't want to be locked out by an attack on the port to
+		 * deny service, so we time out the read, the network had better
+		 * deliver promptly
+		 */
+		do {
+			i = select(fd + 1, &readset, NULL, NULL, &timeout);
+			if (pbs_tcp_interrupt)
+				break;
+		}
 #ifdef WIN32
-	while (i == -1 && errno == WSAEINTR);
+		while (i == -1 && errno == WSAEINTR);
```

Comment:
I think this is not ever set errno - since on windows we need to call WSAGetLastError() to get the error code. Besides that, WASEINTR seems to happen only if when a cancel API is called (which we never call).

From MSDN:
WSAEINTR 10004 | Interrupted function call.  A blocking operation was interrupted by a call to WSACancelBlockingCall.
-- | --